src/main/java/hashtable/HashTable.java

```diff
@@ -0,0 +1,125 @@
+
+    /**
+     * Constructs a new hashtable with inserted capacity.
+     * All slots are null by default.
+     * @param capacity initial capacity of hashtable.
+     * @exception IllegalArgumentException if the specified capacity is less or equal zero.
+     */
+    public HashTable (int capacity) {
+        if (capacity <= 0) {
+            throw new IllegalArgumentException("Capacity cannot be less or equal zero!");
+        }
+        array = new ArrayList<>();
+        array.ensureCapacity(capacity);
+        for (int i = 0; i < capacity; i++) {
+            array.add(null);
+        }
+    }
+
+    /**
+     * Inserts pair in a table.
+     * If pair with such key already exists in a slot, the old pair will be replaced with the new one.
+     * Even If slot is not empty, new pair will be added to this slot as well.
+     * @param key the hashtable key
+     * @param value the value
+     */
+    public void put(K key, V value) {
+        LinkedListNode<K, V> node = getNodeForKey(key);
+        if (node != null) {
+            node.value = value; // Update old value with a new.
+        }
+        node = new LinkedListNode<K, V>(key, value);
+        int index = getIndexForKey(key);
+        if (array.get(index) != null) {
+            node.next = array.get(index);
+            node.next.prev = node;
+        }
+        array.set(index, node);
+    }
+
+    /**
+     * Removes a pair with specified key from the table.
+     * @param key the key that needs to be removed.
+     * @exception IllegalArgumentException if a pair with specified key does not exist in table.
+     */
+    public void remove(K key) {
+        LinkedListNode<K, V> node = getNodeForKey(key);
+        if (node == null) {
+            throw new IllegalArgumentException("Value with such key was not found!");
+        }
+        if (node.prev != null) {
+            node.prev.next = node.next;
+        } else {
+            int hashKey = getIndexForKey(key);
+            array.set(hashKey, node.next);
+        }
+        if (node.next != null) {
+            node.next.prev = node.prev;
+        }
+    }
+
+    /**
+     * Returns the value to which the specified key is mapped.
+     * @param key the key whose associated value is to be returned.
+     * @return the value to which the specified key is mapped or null if no value for specified key.
+     * @exception IllegalArgumentException if specified key is null.
+     */
+    public V getValue(K key) {
+        if (key == null) {
+            throw new IllegalArgumentException("Key cannot be null!");
+        }
+        LinkedListNode<K, V> node = getNodeForKey(key);
+        return node == null ? null : node.value;
+    }
+
+    /**
+     * Returns the index in table for specified key.
+     * @param key the key that needs to be converted into index.
+     * @return index in table.
+     * @exception IllegalArgumentException if specified key is null.
+     */
+    public int getIndexForKey(K key) {
```

Comment:
Ideally the number of public methods is kept to the minimum. Methods like this shouldn't be called from outside, so it can be private.